client/src/components/ElementEditor/Element.js

```diff
@@ -215,6 +296,81 @@ class Element extends Component {
     }
   }
 
+  handleSaveButtonClick() {
+    this.setState({
+      ensureFormRendered: true,
+      doSaveElement: true,
+    });
+  }
+
+  handlePublishButtonClick() {
+    this.setState({
+      justClickedPublishButton: true,
+      ensureFormRendered: true,
+    });
+    if (this.props.formDirty) {
+      // Save the element first before publishing, which may trigger validation errors
+      this.props.submitForm();
```

Comment:
No that's inccorect.

Instead we're saving the form first via this.props.submitForm(). That will trigger validation errors via the whole FormSubmission thing server side.  We need to do this before we publish.

The publish button just hits a generic GraphQL endpoint that publishes a dataobject

We're setting justClickedPublishButton: true before this. What this will do is set 

```js
    if (this.state.justClickedPublishButton) {
      this.setState({
        justClickedPublishButton: false,
        doPublishElement: true,
      });
    }
```

In `handleFormSchemaSubmitResponse()` after a save submission without validation errors has occured

Reply:
This is really confusing to me. We set `justClickedPublishButton` to true, which then gets set to false so we can set `doPublishElement` to true, which then goes and does another thing.... this whole flow of saving and publishing feels really convoluted.

It's clear from discussions with you that this is done to work around limitations in the existing code.... but I think we need a writeup somewhere that says "when you click save, this is the flow of events that happens" and  "when you click publish, this is the flow of events that happens" - otherwise we'll have to unpick all of this any time we need to do maintenance in this area.

It's also really hard to review this code without something that tells me how it's all wired together - it's too hard to trace it all.